src/analyses/goto_check.cpp

```diff
@@ -1226,21 +1230,21 @@ void goto_checkt::pointer_primitive_check(
                        : size_of_expr_opt.value();
 
   const conditionst &conditions = address_check(pointer, size);
-
-  exprt::operandst conjuncts;
-
   for(const auto &c : conditions)
-    conjuncts.push_back(c.assertion);
-
-  const or_exprt or_expr(null_object(pointer), conjunction(conjuncts));
-
-  add_guarded_property(
-    or_expr,
-    "pointer in pointer primitive is neither null nor valid",
-    "pointer primitives",
-    expr.source_location(),
-    expr,
-    guard);
+  {
+    // we want all address checks *except* for the null check because these
+    // should only fail if the assumption is on in invalid non-null pointer.
+    if(c.description != pointer_is_null_check_description)
+    {
+      add_guarded_property(
+        or_exprt{null_object(pointer), c.assertion},
```

Comment:
:no_entry_sign: This doesn't actually appear to be adding the equivalent checks in the case where `mode == ID_java` or where `flags.is_null()` is false.

There must be a way to implement this which is both more readable and more straight forward to get correct. Besides the issue of the string comparison not being great, I think that filtering the list of conditions after generating them is fundamentally a sub-optimal approach. I would suggest starting by looking at splitting the `goto_checkt::address_check` function into multiple functions. At least one of which should generate the list of conditions without the null check. That way the check never has to be filtered out after being unnecessarily added.